File: cpp/lib/testcontainers/Container.cpp
#include <rust/cxx.h>
#include <rust_tc_bridge/lib.h>

#include "testcontainers/Container.hpp"
#include "testcontainers/core/ExecCommand.hpp"
#include "testcontainers/core/SyncExecResult.hpp"
#include "testcontainers/system/UrlHost.hpp"
#include "details/OptionHelper.hpp"

#include "details/BoxHelper.hpp"
#include "details/ErrorHelper.hpp"

namespace testcontainers {

Container::Container(RsContainer *container) noexcept
    : rimpl_(container, [](RsContainer *c) { ::rs_container_destroy(details::box_from_raw(c)); }) {}

Container::Container(Container &&other) noexcept = default;

Container &Container::operator=(Container &&other) noexcept = default;

Container::~Container() noexcept = default;

bool Container::is_valid() const noexcept { return static_cast<bool>(rimpl_); }

std::uint16_t Container::get_host_port_ipv4(ContainerPort port) const {
  return details::call_map_error(&RsContainer::rs_container_get_host_port_ipv4, rimpl_.get(),
                                 details::into_box(port.rimpl_));
}

std::uint16_t Container::get_host_port_ipv6(ContainerPort port) const {
  return details::call_map_error(&RsContainer::rs_container_get_host_port_ipv6, rimpl_.get(),
                                 details::into_box(port.rimpl_));
}

UrlHost Container::get_host() const {
  return UrlHost(details::call_map_error(&RsContainer::rs_container_get_host, rimpl_.get()).into_raw());
}

SyncExecResult Container::exec(ExecCommand cmd) const {
  return SyncExecResult(
      details::call_map_error(&RsContainer::rs_container_exec, rimpl_.get(), details::into_box(cmd.rimpl_))
          .into_raw());
}

void Container::stop() const { details::call_map_error(&RsContainer::rs_container_stop, rimpl_.get()); }

void Container::stop_with_timeout(std::optional<std::int32_t> timeout_sec) const {
  details::call_map_error(&RsContainer::rs_container_stop_with_timeout, rimpl_.get(),
                           utils::optional_to_vec(std::move(timeout_sec)));
}

void Container::start() const { details::call_map_error(&RsContainer::rs_container_start, rimpl_.get()); }

void Container::rm(Container container) {
  details::call_map_error(::rs_container_rm, details::into_box(container.rimpl_));
}

std::vector<std::uint8_t> Container::stdout_to_vec() const noexcept {
  auto rust_vec = rimpl_->rs_container_stdout_to_vec();
  return std::vector<std::uint8_t>(rust_vec.begin(), rust_vec.end());
}

std::vector<std::uint8_t> Container::stderr_to_vec() const noexcept {
  auto rust_vec = rimpl_->rs_container_stderr_to_vec();
  return std::vector<std::uint8_t>(rust_vec.begin(), rust_vec.end());
}

bool Container::is_running() const noexcept { return rimpl_->rs_container_is_running(); }

std::optional<std::int64_t> Container::exit_code() const noexcept {
  return utils::vec_to_optional(rimpl_->rs_container_exit_code_opt());
}

std::string Container::stdout_to_string() const {
  const auto out = stdout_to_vec();
  return std::string(out.begin(), out.end());
}

std::string Container::stderr_to_string() const {
  const auto err = stderr_to_vec();
  return std::string(err.begin(), err.end());
}

} // namespace testcontainers

File: rust/src/container.rs
use crate::{
    core::exec::exec_command::RsExecCommand, core::exec::sync_exec_result::RsSyncExecResult,
    image::RsGenericImage, system::ip::ip_addr::RsIpAddr, system::url_host::RsUrlHost,
    core::container_port::RsContainerPort,
};
use testcontainers::{core::ContainerPort, Container, GenericImage};

pub struct RsContainer {
    container: Container<GenericImage>,
}

pub fn rs_container_destroy(container: Box<RsContainer>) {
    drop(container);
}

pub fn rs_container_rm(container: Box<RsContainer>) -> Result<(), String> {
    container
        .container
        .rm()
        .map_err(|e| format!("Failed to remove container: {}", e))
}

impl RsContainer {
    pub fn new(container: Container<GenericImage>) -> Self {
        Self { container }
    }

    pub fn rs_container_id(self: &RsContainer) -> &str {
        self.container.id()
    }

    pub fn rs_container_image(self: &RsContainer) -> Box<RsGenericImage> {
        Box::new(RsGenericImage::new(self.container.image().clone()))
    }

    // TODO HashMap<T,U>
    // pub fn ports(&self) -> Result<Ports>

    pub fn rs_container_get_host_port_ipv4(
        self: &RsContainer,
        port: Box<RsContainerPort>,
    ) -> Result<u16, String> {
        let port: ContainerPort = (*port).into();
        self.container
            .get_host_port_ipv4(port)
            .map_err(|e| format!("Failed to get host port: {}", e))
    }

    pub fn rs_container_get_host_port_ipv6(
        self: &RsContainer,
        port: Box<RsContainerPort>,
    ) -> Result<u16, String> {
        let port: ContainerPort = (*port).into();
        self.container
            .get_host_port_ipv6(port)
            .map_err(|e| format!("Failed to get host port: {}", e))
    }

    pub fn rs_container_get_bridge_ip_address(self: &RsContainer) -> Result<Box<RsIpAddr>, String> {
        self.container
            .get_bridge_ip_address()
            .map_err(|e| format!("Failed to get bridge IP address: {}", e))
            .map(|ip_addr| Box::new(RsIpAddr::new(ip_addr)))
    }

    pub fn rs_container_get_host(self: &RsContainer) -> Result<Box<RsUrlHost>, String> {
        self.container
            .get_host()
            .map_err(|e| format!("Failed to get host: {}", e))
            .map(|host| Box::new(RsUrlHost::new(host)))
    }

    pub fn rs_container_exec(
        self: &RsContainer,
        cmd: Box<RsExecCommand>,
    ) -> Result<Box<RsSyncExecResult>, String> {
        self.container
            .exec(cmd.command)
            .map_err(|e| format!("Failed to exec command: {}", e))
            .map(|exec_result| Box::new(RsSyncExecResult::new(exec_result)))
    }

    pub fn rs_container_stop(self: &RsContainer) -> Result<(), String> {
        self.container
            .stop()
            .map_err(|e| format!("Failed to stop container: {}", e))
    }

    pub fn rs_container_stop_with_timeout(
        self: &RsContainer,
        timeout_sec_opt: Vec<i32>,
    ) -> Result<(), String> {
        self.container
            .stop_with_timeout(timeout_sec_opt.first().copied())
            .map_err(|e| format!("Failed to stop container with timeout: {}", e))
    }

    pub fn rs_container_start(self: &RsContainer) -> Result<(), String> {
        self.container
            .start()
            .map_err(|e| format!("Failed to start container: {}", e))
    }

    // TODO async
    // pub async fn rs_container_pause(self: &RsContainer) -> Result<(), String>
    // pub async fn rs_container_unpause(self: &RsContainer) -> Result<(), String>

    // TODO callbacks
    // pub fn stdout(&self, follow: bool) -> Box<dyn BufRead + Send>
    // pub fn stderr(&self, follow: bool) -> Box<dyn BufRead + Send>

    pub fn rs_container_stdout_to_vec(self: &RsContainer) -> Result<Vec<u8>, String> {
        self.container
            .stdout_to_vec()
            .map_err(|e| format!("Failed to get stdout: {}", e))
    }

    pub fn rs_container_stderr_to_vec(self: &RsContainer) -> Result<Vec<u8>, String> {
        self.container
            .stderr_to_vec()
            .map_err(|e| format!("Failed to get stderr: {}", e))
    }

    pub fn rs_container_is_running(self: &RsContainer) -> Result<bool, String> {
        self.container
            .is_running()
            .map_err(|e| format!("Failed to get is running: {}", e))
    }

    pub fn rs_container_exit_code_opt(self: &RsContainer) -> Result<Vec<i64>, String> {
        self.container
            .exit_code()
            .map_err(|e| format!("Failed to get exit code: {}", e))
            .map(|opt| opt.into_iter().collect())
    }

    // TODO async
    // fn rt(&self) -> &Arc<tokio::runtime::Runtime>
    // fn async_impl(&self) -> &ContainerAsync<I>
}
